pkg/security/ebpf/c/raw_syscalls.h

```diff
@@ -53,31 +44,103 @@ int sys_enter(struct _tracepoint_raw_syscalls_sys_enter *args) {
     bpf_probe_read(&syscall.id, sizeof(syscall.id), &args->id);
     bpf_get_current_comm(&syscall.comm, sizeof(syscall.comm));
 
-    u32 zero = 0;
-    u32 *back_buffer;
-    if (!(back_buffer = bpf_map_lookup_elem(&noisy_processes_buffer, (void*) &zero)))
+    u32 selector_key = SYSCALL_MONITOR_KEY;
+    u32 *buffer_id;
+    if (!(buffer_id = bpf_map_lookup_elem(&buffer_selector, &selector_key)))
         return 0;
 
-    u64 count = 0;
-    u64 *current_count;
-
-    if (*back_buffer) {
-        current_count = bpf_map_lookup_elem(&noisy_processes_bb, &syscall);
+    u64 *count;
+    if (*buffer_id) {
+        count = bpf_map_lookup_elem(&noisy_processes_bb, &syscall);
     } else {
-        current_count = bpf_map_lookup_elem(&noisy_processes_fb, &syscall);
+        count = bpf_map_lookup_elem(&noisy_processes_fb, &syscall);
     }
 
-    if (current_count) {
-        count = *current_count;
+    if (count) {
+        __sync_fetch_and_add(count, 1);
+    } else {
+        u64 one = 1;
+        count = &one;
+
+        if (*buffer_id) {
+            bpf_map_update_elem(&noisy_processes_bb, &syscall, count, BPF_ANY);
+        } else {
+            bpf_map_update_elem(&noisy_processes_fb, &syscall, count, BPF_ANY);
+        }
     }
-    count++;
 
-    if (*back_buffer) {
-        bpf_map_update_elem(&noisy_processes_bb, &syscall, &count, BPF_ANY);
+    return 0;
+}
```

Comment:
It could be interesting to extract some plumbing logic that is shared between `sys_enter` and `sched_process_exec` into dedicated utility functions.
I can see two “tasks” that are repeated and that are generic enough to be extracted:
* The selection of the right map to update according to the value of the `buffer_selector` map.
* The initialisation of a map element if it doesn’t exist. (Basically, the job done by the [`lookup_or_try_init` bcc function](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#18-maplookup_or_try_init).

Something like this:

```c
static __attribute__((always_inline))
struct bpf_map_def *select_buffer(struct bpf_map_def *front_buffer,
                                  struct bpf_map_def *back_buffer) {
    u32 selector_key = SYSCALL_MONITOR_KEY;
    u32 *buffer_id = bpf_map_lookup_elem(&buffer_selector, &selector_key)
    if (buffer_id == NULL)
        return NULL;

    return *buffer_id ? back_buffer : front_buffer;
}
```

```c
static __attribute__((always_inline))
void *bpf_map_lookup_or_try_init(struct bpf_map_def *map, void *key, void *zero) {
    void *value = bpf_map_lookup_elem(map, key);
    if (value != NULL)
        return value;

    // Use BPF_NOEXIST to prevent race condition
    if (bpf_map_update_elem(map, key, zero, BPF_NOEXIST) < 0)
        return NULL;

    return bpf_map_lookup_elem(map, key);
}
```

Thanks to the two above utility functions, `sys_enter` can be simplified:
```c
SEC("tracepoint/raw_syscalls/sys_enter")
int sys_enter(struct _tracepoint_raw_syscalls_sys_enter *args) {
    struct process_syscall_t syscall = {};
    bpf_probe_read(&syscall.pid, sizeof(syscall.pid), &args->common_pid);
    bpf_probe_read(&syscall.id, sizeof(syscall.id), &args->id);
    bpf_get_current_comm(&syscall.comm, sizeof(syscall.comm));
    u64 zero = 0;

    struct bpf_map_def *noisy_processes = select_buffer(&noisy_processes_fb, &noisy_processes_bb);
    if (noisy_processes == NULL)
        return 0;

    u64 *count = bpf_map_lookup_or_try_init(noisy_processes, &syscall, &zero);
    if (count == NULL)
        return 0;

    __sync_fetch_and_add(count, 1);

    return 0;
```